app/src/main/java/tech/ula/utils/ValidationUtility.kt

```diff
@@ -1,31 +1,86 @@
+        var errorMessage = ""
+        val blacklist = resources.getStringArray(R.array.blacklisted_usernames)
 
-        pattern = Pattern.compile(USERNAME_REGEX)
-        matcher = pattern.matcher(password)
+        when {
+            username.isEmpty() -> {
+                errorMessage = resources.getString(R.string.error_empty_field)
+            }
+            !validateUsernameCharacters(username) -> {
+                errorMessage = resources.getString(R.string.error_username_invalid_characters)
+            }
+            blacklist.contains(username) -> {
+                errorMessage = resources.getString(R.string.error_username_in_blacklist)
+                errorMessage = String.format(errorMessage, username)
+            }
+            else ->
+                usernameIsValid = true
+        }
 
-        return matcher.matches()
+        return Credential(usernameIsValid, errorMessage)
     }
 
-    fun isPasswordValid(password: String): Boolean {
+    fun validatePasswords(password: String, vncPassword: String): Credential {
+        var passwordIsValid = false
+        var errorMessage = ""
+
+        when {
+            password.isEmpty() || vncPassword.isEmpty() -> {
+                errorMessage = resources.getString(R.string.error_empty_field)
+            }
+            vncPassword.length > 8 || vncPassword.length < 6 -> {
+                errorMessage = resources.getString(R.string.error_vnc_password_length_incorrect)
+            }
+            !validatePasswordCharacters(password) -> {
+                errorMessage = resources.getString(R.string.error_password_invalid)
+            }
+            !validatePasswordCharacters(vncPassword) -> {
+                errorMessage = resources.getString(R.string.error_vnc_password_invalid)
+            }
+            else ->
+                passwordIsValid = true
+        }
+
+        return Credential(passwordIsValid, errorMessage)
+    }
+
+    private fun validateUsernameCharacters(username: String): Boolean {
+        val usernameRegex = "([a-z_][a-z0-9_]{0,30})"
+
+        val compiledRegex = Pattern.compile(usernameRegex)
+        val matcher = compiledRegex.matcher(username)
+        val hasValidCharacters = matcher.matches()
+
+        if (hasValidCharacters) {
+            return true
+        }
+        return false
+    }
+
+    private fun validatePasswordCharacters(password: String): Boolean {
         val pattern: Pattern
         val matcher: Matcher
 
-        val PASSWORD_REGEX = "^[a-zA-Z0-9!@#$%^&*()_+=,./?<>:]*\$"
+        val passwordRegex = "^[a-zA-Z0-9!@#$%^&*()_+=,./?<>:]*\$"
 
-        pattern = Pattern.compile(PASSWORD_REGEX)
+        pattern = Pattern.compile(passwordRegex)
         matcher = pattern.matcher(password)
 
         return matcher.matches()
     }
-}
\ No newline at end of file
+}
+
+sealed class CredentialValidation
+data class Credential(val credentialIsValid: Boolean, val errorMessage: String) : CredentialValidation()
```

Comment:
It looks like right now, we're only using the `errorMessage` field to populate toasts. As an FYI, toasts also work with resource ids, which means we could just pass those back. Resource ids are under the app specific `R` package, which means you can import them into the `ValidationUtility`, but you could also access it in the tests, which would save you from needing to assert that the strings are equal (you could just assert the id is the expected). It would also save you from needing to inject a `Resources` object into the utility.

I don't care enough that I think you actually need to change it, because it seems like a fair amount of work for not a ton of benefit, and it's also easy to imagine a scenario where you'd want the error as a string.

Reply:
Good point, I've got it now so that it will compare resource Ids rather than strings, which eliminates the need for injecting the `Resource` object